pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,203 @@
+		registry: registry,
+		nodeID:   "",
+	}
+
+	return &le, nil
+}
+
+// LeaderElector is a leader election client.
+type LeaderElector struct {
+	// nodeID represents the id assigned by registry
+	nodeID string
+
+	// leaderID stores id the current leader
+	leaderID string
+
+	// Leader election config
+	config Config
+
+	// registry being used for leader election
+	registry registry.Registry
+}
+
+// Run starts the leader election loop. Run will not return
+// before leader election loop is stopped by ctx or it has
+// stopped holding the leader lease
+func (le *LeaderElector) Run(ctx context.Context) {
+	defer runtime.HandleCrash()
+	defer func() {
+		le.config.Callbacks.OnStoppedLeading()
+	}()
+
+	if !le.acquire(ctx) {
+		return // ctx signalled done
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go le.config.Callbacks.OnStartedLeading(ctx)
+	le.renew(ctx)
+}
+
+// GetLeader returns the identity of the last observed leader or returns the empty string if
+// no leader has yet been observed.
+func (le *LeaderElector) GetLeader() string {
+	return le.nodeID
+}
+
+// IsLeader returns true if the last observed leader was this client else returns false.
+func (le *LeaderElector) IsLeader() bool {
+	// TODO: fetch latest data from registry for current leader
+	return (le.nodeID != "") && (le.nodeID == le.leaderID)
+}
+
+// acquire loops calling tryAcquire and returns true immediately when tryAcquire succeeds.
+// Returns false if ctx signals done.
+func (le *LeaderElector) acquire(ctx context.Context) bool {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	succeeded := false
+
+	logger.Log.Info("attempting to acquire leader lease")
+	wait.JitterUntil(func() {
+		succeeded = le.tryAcquireOrRenew(ctx)
+		//le.maybeReportTransition()
+		if !succeeded {
+			logger.Log.Info("failed to acquire lease")
+			return
+		}
+
+		logger.Log.Info("successfully acquired lease")
+		cancel()
+	}, le.config.RetryPeriod, JitterFactor, true, ctx.Done())
+	return succeeded
+}
+
+// tryAcquire tries to acquire a leader lease if it is not already acquired
+// Returns true on success else returns false.
+func (le *LeaderElector) tryAcquireOrRenew(ctx context.Context) bool {
+	var err error
+
+	// validate if node is already registered
+	if le.nodeID != "" {
+		// if node is not registred anymore, we will need register it again
+		// it can happen that ttl associated with registration had expired
+		if ok := le.registry.IsRegistered(le.nodeID); !ok {
+			logger.Log.Info("node is unregisterd")
+			le.nodeID = ""
+		} else {
+			err = le.registry.Renew(le.nodeID)
+			if err != nil {
+				logger.Log.Infof("error while renewing registration: %v", err.Error())
+				return false
+			}
+		}
+	}
+
+	// if not registered, register it
+	if le.nodeID == "" {
+		le.nodeID, err = le.registry.Register(le.config.Name, le.config.LeaseDuration)
+		if err != nil {
+			logger.Log.Errorf("failed to register node with registry: %v", err)
+			return false
```

Comment:
should bool be returned only when leader is acquired or not. In case of error in registering or renewing lease, shouldn't we return error and let it be handled at upper layers where the consumer can even shutdown?

Reply:
There are 2 infinite loops running outside
1. Acquiring - This cares only about acq and retry on all failures
2. Renew - This cares only about renew success and retry for all failures

hence kept it same way, otherwise upper layer will need to handle both err and bool but action will be same to retry

Reply:
But infinite retry is also problematic. Lets discuss this